File: 03 Python/azimuth_elevation/tle_utils.py
import requests

def get_tles_from_url(url):
  tleRes = requests.get(url)
  tleTxt = tleRes.text
  tleLines = tleTxt.split("\n")

  tles = []

  for idx in range(0, len(tleLines) - 2, 3):
    tles.append({
      "name": tleLines[idx + 0].strip(),
      "tle": [
        tleLines[idx + 1].strip(),
        tleLines[idx + 2].strip(),
      ]
    })

  return tles

def fetch_tles(root=".", groups=None):
  if len(root) < 1:
    root = "."

  if groups is None or len(groups) < 1:
    groups = ["goes", "noaa", "starlink", "kuiper", "sss", "sci", "gps", "glonass", "galileo", "beidou"]

  urls = { k: f"{root}/data/tles/{k}.txt" for k in groups }
  data = { k: get_tles_from_url(u) for k,u in urls.items() }

  return data

File: 02 JavaScript/azimuth-elevation/tleUtils.js
async function getTleFromUrl(url) {
  const tleRes = await fetch(url);
  const tleTxt = await tleRes.text();
  const tleLines = tleTxt.split("\n");

  const tles = [];

  for (let ln = 0; ln < tleLines.length - 2; ln += 3) {
    tles.push({
      name: tleLines[ln + 0].trim(),
      tle: [
        tleLines[ln + 1].trim(),
        tleLines[ln + 2].trim(),
      ]
    });
  }
  return tles;
}

async function fetchTles(root = ".", satGroups = []) {
  if (root.length < 1) {
    root = ".";
  }

  if (satGroups.length < 1) {
    satGroups = ["goes", "noaa", "starlink", "kuiper", "sss", "sci", "gps", "glonass", "galileo", "beidou"];
  }

  // { k: f"../data/etc/{k}.txt" for k in satGroups }
  const URLS = satGroups.reduce((acc, k) => ({ ...acc, [k]: `${root}/data/tles/${k}.txt` }), {});

  const ps = satGroups.map(k => getTleFromUrl(URLS[k]));
  const data = await Promise.all(ps);

  return satGroups.reduce((acc, k, idx) => ({ ...acc, [k]: data[idx] }), {});
}

export { getTleFromUrl, fetchTles };
